gprofiler/metadata/application_identifiers.py

```diff
@@ -0,0 +1,256 @@
+
+
+class _ApplicationIdentifier(metaclass=ABCMeta):
+    @abstractmethod
+    def get_application_name(self, process: Process) -> Optional[str]:
+        pass
+
+
+class _GunicornApplicationIdentifier(_ApplicationIdentifier):
+    def get_application_name(self, process: Process) -> Optional[str]:
+        # As of gunicorn documentation the WSGI module name most probably will come from the cmdline and not from the
+        # config file / environment variables (they added the option to specify `wsgi_app`
+        # in the config file only in version 20.1.0)
+
+        if "gunicorn" != os.path.basename(
+            _get_cli_arg_by_index(process.cmdline(), 0)
+        ) and "gunicorn" != os.path.basename(_get_cli_arg_by_index(process.cmdline(), 1)):
+            return None
+
+        # wsgi app specification will come always as the last argument (if hasn't been specified config file)
+        wsgi_app_spec = process.cmdline()[-1].split(":", maxsplit=1)[0]
+        return f"gunicorn: {_append_python_module_to_proc_wd(process, wsgi_app_spec)}"
+
+
+class _UwsgiApplicationIdentifier(_ApplicationIdentifier):
+    @staticmethod
+    def _find_wsgi_from_config_file(process: Process) -> Optional[str]:
+        for arg in process.cmdline():
+            if arg.endswith(".ini"):
+                config_file = arg
+                break
+        else:
+            return None
+
+        if not os.path.isabs(config_file):
+            config_file = os.path.join(process.cwd(), config_file)
+
+        config = configparser.ConfigParser()
+        config.read(resolve_host_path(process, config_file))
+        try:
+            return config.get("uwsgi", "module")
+        except (configparser.NoSectionError, configparser.NoOptionError):
+            pass
+
+        return None
+
+    def get_application_name(self, process: Process) -> Optional[str]:
+        if "uwsgi" != os.path.basename(_get_cli_arg_by_index(process.cmdline(), 0)):
+            return None
+
+        wsgi_arg = _get_cli_arg_by_name(process.cmdline(), "-w") or _get_cli_arg_by_name(
+            process.cmdline(), "--wsgi-file"
+        )
+        if wsgi_arg is not _NON_AVAILABLE_ARG:
+            return f"uwsgi: {_append_python_module_to_proc_wd(process, wsgi_arg)}"
+
+        wsgi_config = self._find_wsgi_from_config_file(process)
+        if wsgi_config is not None:
+            return f"uwsgi: {_append_python_module_to_proc_wd(process, wsgi_config)}"
+
+        _logger.warning(
+            f"{self.__class__.__name__} Couldn't find uwsgi wsgi module, both from cmdline and from config file",
+            cmdline=process.cmdline(),
+            no_extra_to_server=True,
+        )
+        return None
+
+
+class _CeleryApplicationIdentifier(_ApplicationIdentifier):
+    @staticmethod
+    def is_celery_process(process: Process) -> bool:
+        if "celery" == os.path.basename(_get_cli_arg_by_index(process.cmdline(), 0)) or "celery" == os.path.basename(
```

Comment:
For celery you check argv0, argv1. For uwsgi you check only argv0. Is it never `python /path/to/uwsgi/uwsgi`? uwsgi is a binary compiled w/ cpython, yeah?

Reply:
yep

Reply:
k